### ISSUE TITLE
geth preload option not working?

### ISSUE THREAD

**Issue by Paulohvk** [ethereum/go-ethereum#24125]
Hi,

apologies for cross posting, but I am having the same issue as in [this thread](https://github.com/ethereum/go-ethereum/issues/21190), which hasn't been resolved.

#### System information

Geth version: `geth version` 1.10.9
Consensus: Clique PoA
OS & Version: Ubuntu 20.04

#### Expected behaviour
Basically, I have a private geth network with 2 nodes (only 1 miner) connected to each other. However, recently I noticed that my miner is not mining automatically anymore (I used to use the script provided in [this thread](https://ethereum.stackexchange.com/questions/3151/how-to-make-miner-to-mine-only-when-there-are-pending-transactions/3152#3152) to mine only non-empty blocks) by doing:

`geth --datadir ./Node1/ --preload './mineWhenNeeded.js'`

However, it seems that geth is not able to load the script, as whenever I perform a transaction through the geth console, the transaction shows up in txpool.status as pending and the eth.blockNumber never goes up.

(I have also tried different paths to my script, which also do not work...)

#### Actual behaviour

However, if I attach to the geth console and load the script manually through loadScript then the script works fine and mines the pending block and stops mining afterwards.

I use the following command in the geth console to load the script:

`loadScript('./mineWhenNeeded.js')`

#### Steps to reproduce the behaviour
1 - Download and save the script provided [here](https://ethereum.stackexchange.com/questions/3151/how-to-make-miner-to-mine-only-when-there-are-pending-transactions/3152#3152).
2 - Initialize the geth node by including the preload command trying to load the specified script.
3 - Attach a geth console to the node.
4 - Unlock account and send a transaction.
5 - Verify that the script is not being loaded, as the transaction is eternally pending.
6 - Manually load the script by using loadScript('./mineWhenNeeded.js'), which works as intended or manually start mining with miner.start(1) then miner.stop() to send transaction.

Any thoughts on this?

Thanks in advance!

**Comment by s1na** [comment]
So the `--preload` flag is to be used as part of one of these commands: `console`, `attach`, `js`. I.e. you'd need to run:

```sh
geth --datadir ./Node1/ --preload './mineWhenNeeded.js' console
```

That's because by default running geth doesn't spin up a JS engine.

**Comment by Paulohvk** [comment]
Hi,

Thank you very much for the feedback!

Unfortunately by using the preload option followed by the console command, it still does not work, it just opens a console in the node, so I would have to manually load the script by using the loadScript command.

However, by trying your other suggestions I managed to fix the issue, by using the following:

`geth --datadir ./Node1/ js './mineWhenNeeded.js'`

This way, I avoid using the preload command (which still does not work), and use the js command to specify which script to run.

Many thanks again!

**Comment by holiman** [comment]
```
$ geth --dev  --preload foo.js console
INFO [01-20|10:52:49.186] Starting Geth in ephemeral dev mode... 
...
INFO [01-20|10:52:49.441] Commit new mining work                   number=1 sealhash=b25d27..006aab uncles=0 txs=0 gas=0 fees=0 elapsed="109.096µs"
WARN [01-20|10:52:49.441] Block sealing failed                     err="sealing paused while waiting for transactions"
test
Welcome to the Geth JavaScript console!

instance: Geth/v1.10.16-unstable-b1f09596-20220116/linux-amd64/go1.17.3
coinbase: 0x33afe373bff5ef199d4e6ad1426281723fa4471d
at block: 0 (Thu Jan 01 1970 01:00:00 GMT+0100 (CET))

```
```
$ cat foo.js
console.log("test")
```
Works fine for me. Please reopen if I've missed something

**Comment by holiman** [comment]
And with a datadir somewhere else: 
```
$ geth --nodiscover --maxpeers 0  --datadir=/tmp/foo --preload ./foo.js console
INFO [01-20|10:55:10.394] Starting Geth on Ethereum mainnet... 
...

INFO [01-20|10:55:10.706] New local node record                    seq=1,642,672,492,100 id=02a31fb57a600e77 ip=127.0.0.1 udp=0 tcp=30303
INFO [01-20|10:55:10.706] Started P2P networking                   self="enode://3c2de9d51e8dd4df460eac99e2f7d29eb60ceb680ada88c50e4aebd3e4eb5a278d636bd7231930e8de81f614f5544d99ef969ac83d216da59371732c77349093@127.0.0.1:30303?discport=0"
INFO [01-20|10:55:10.707] IPC endpoint opened                      url=/tmp/foo/geth.ipc
WARN [01-20|10:55:10.708] Disk space is running low. Geth will shutdown if disk space runs below critical level. available=1009.14MiB critical_level=887.00MiB
test
```
